src/librustdoc/html/render/mod.rs

```diff
@@ -1527,7 +1510,7 @@ fn render_impl(
             );
         }
     }
-    let details_str = if impl_items.is_empty() {
+    let details_str = if impl_items.is_empty() && default_impl_items.is_empty() {
```

Comment:
I have an idea for how to make this a little clearer / safer:

```
let open_details = || {
    if !impl_items.is_empty() || default_impl_items.is_empty() {
      write!(w, details_str);
      close_tags.insert_str(0, "</details>");
    }
};
```

Then, in each place below where you interpolate `details_str`, you would instead call `open_details()`. That way we ensure that if the details tag is opened, it's also closed, without having to duplicate the `if` logic a bunch of times.

Reply:
I have a few ideas on how to improve it even further. You'll tell me what you think of it. :)